src/CommandBasedBackend.cpp

```diff
@@ -148,4 +217,148 @@ namespace ChimeraTK {
+   * Throws a ChimeraTK::logic_error if the json key is not in an std::array of valid keys
+   */
+  template<size_t N>
+  void throwIfHasInvalidJsonKey(const json& j, const std::array<std::string, N>& validKeys, std::string errorMessage) {
+    for(auto it = j.begin(); it != j.end(); ++it) {
+      bool stringIsNotInArray = true;
+      for(const std::string& i : validKeys) {
+        if(it.key() == i) {
+          stringIsNotInArray = false;
+          break;
+        }
+      }
+
+      if(stringIsNotInArray) {
+        throw ChimeraTK::logic_error(errorMessage + " " + it.key());
+      }
+    }
+  }
+
+  /******************************************************************************************************************/
+
+  void CommandBasedBackend::parseJsonAndPopulateCatalogue(const std::string& mapFileName) {
+    std::ifstream file(mapFileName);
+    if(!file.is_open()) {
+      throw ChimeraTK::logic_error("Could not open the map file " + mapFileName);
+    }
+
+    json j;
+    file >> j;
+
+    if(!j.contains(to_str(METADATA)) || !j.contains(to_str(REGISTERS))) {
+      throw ChimeraTK::logic_error("Missing keys in JSON data");
+    }
+
+    throwIfHasInvalidJsonKey(j, topLevelKeyStrs, "Map file top level has unknown key");
+
+    _metaData = j.at(to_str(METADATA)).template get<CommandBasedBackend::MetaData>();
+    // get() calls from_json(const json& j, CommandBasedBackend::MetaData& m)
+
+    for(auto& [key, value] : j.at(to_str(REGISTERS)).items()) {
+      throwIfHasInvalidJsonKey(value, registerKeyStrs, "Map file registry entry " + key + " has unknown key");
+
+      CommandBasedBackendRegisterInfo reg = getJsonToCommandBasedBackendRegisterInfo(value, key);
+      reg.registerPath = RegisterPath(key);
+      _backendCatalogue.addRegister(reg);
+    }
+  }
+
+  /********************************************************************************************************************/
+
+  void from_json(const json& j, CommandBasedBackend::MetaData& m) {
+    if(!j.contains(to_str(MAP_FILE_FORMAT_VERSION))) {
+      throw ChimeraTK::logic_error("Missing mapFileFormatVersion key in metadata");
+    }
+
+    throwIfHasInvalidJsonKey(j, metadataKeyStrs, "Map file metadata has unknown key");
+
+    if(int mapFileFormatVersion = j.value(to_str(MAP_FILE_FORMAT_VERSION), 0);
+        mapFileFormatVersion != requiredMapFileFormatVersion) {
+      throw ChimeraTK::logic_error("Incorrect map file format version " + std::to_string(mapFileFormatVersion) +
+          ", version " + std::to_string(requiredMapFileFormatVersion) + " required.");
+    }
+
+    m.defaultRecoveryRegister = RegisterPath(j.value(to_str(DEFAULT_RECOVERY_REGISTER), ""));
+
+    j.at(to_str(DELIMITER)).get_to(m.serialDelimiter);
+  }
+
+  /******************************************************************************************************************/
+
+  CommandBasedBackendRegisterInfo getJsonToCommandBasedBackendRegisterInfo(const json& j, const std::string& regKey) {
```

Comment:
The name is not good. It mixes getSomething() with doSomething(). what exactly do you get? JsonTo ?

parseRegisterInfo()? registerInfoFromJson()? jsonToRegisterInfo()? That it's command based can be neglected here because this is a local function which only exists here.

Reply:
But again, it's not super critical because the name is purely local here.

Reply:
I'll go with registerInfoFromJson to ease the read order.